File: iOS/VeLiveQuickStartDemo/VeLiveQuickStartDemo/Helper/File/VeLiveFileReader.m
/**
 * Copyright (c) 2023 Beijing Volcano Engine Technology Ltd. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
//
// VeLiveFileReader.m
//
//  Created by Volcano Engine Team on 2024/11/21.
//

#import "VeLiveFileReader.h"

@interface VeLiveFileReader ()
@property (nonatomic, strong) dispatch_source_t timer;
@property (atomic, copy) VELFileDataBlock dataBlock;
@property (atomic, copy) FLEFileReadCompletionBlock completionBlock;
@property (nonatomic, strong) VeLiveFileConfig *fileConfig;
@property (atomic, assign) BOOL isPaused;
@property (atomic, assign) BOOL isStoped;
@property (atomic, strong) NSInputStream *inputStream;
@property (nonatomic, strong) dispatch_queue_t timerQueue;
@end

@implementation VeLiveFileReader
+ (instancetype)readerWithConfig:(__kindof VeLiveFileConfig *)config {
    VeLiveFileReader *reader = [[self alloc] init];
    reader.fileConfig = config;
    return reader;
}
- (void)startWithDataCallBack:(VELFileDataBlock)dataCallBack completion:(FLEFileReadCompletionBlock)completion {
    self.isStoped = NO;
    self.isPaused = NO;
    self.dataBlock = dataCallBack;
    self.completionBlock = completion;
    
    if (![self prepareFileRead]) {
        return;
    }
    
    if (!_timer) {
        NSString *queueName = [NSString stringWithFormat:@"com.velive.filereader.%@", self.fileConfig.path.lastPathComponent];
        self.timerQueue = dispatch_queue_create([queueName cStringUsingEncoding:NSUTF8StringEncoding], DISPATCH_QUEUE_SERIAL);
        _timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, self.timerQueue);
        dispatch_resume(_timer);
    }
    
    NSTimeInterval interval = self.fileConfig.interval;
    dispatch_source_set_timer(_timer, dispatch_time(DISPATCH_TIME_NOW, 0), interval * NSEC_PER_SEC, 0.01 * NSEC_PER_SEC);
    __weak __typeof__(self)weakSelf = self;
    dispatch_source_set_event_handler(_timer, ^{
        __strong __typeof__(weakSelf)self = weakSelf;
        [self readFileAndCallBack];
    });
}

- (void)stop {
    self.isStoped = YES;
    if (_timer != nil) {
        dispatch_source_cancel(_timer);
        _timer = nil;
    }
}

- (void)pause {
    self.isPaused = YES;
}

- (void)resume {
    self.isPaused = NO;
}

- (BOOL)prepareFileRead {
    if (!self.fileConfig.isValid) {
        if (self.completionBlock) {
            self.completionBlock(VEL_ERROR(-1, @"文件格式错误或者不存在"), NO);
        }
        return NO;
    }
    if (self.inputStream == nil || !self.inputStream.hasBytesAvailable) {
        [self.inputStream close];
        self.inputStream = [NSInputStream inputStreamWithFileAtPath:self.fileConfig.path];
        [self.inputStream open];
    }
    return YES;
}

- (void)readFileAndCallBack {
    if (self.isPaused || self.isStoped || self.dataBlock == nil) {
        return;
    }
    if (!self.inputStream.hasBytesAvailable) {
        if (self.repeat) {
            [self prepareFileRead];
        } else {
            if (self.completionBlock) {
                vel_sync_main_queue(^{
                    self.completionBlock(nil, YES);
                });
            }
        }
        return;
    }
    
    int packetSize = self.fileConfig.packetSize;
    uint8_t *buffer = (uint8_t *)malloc(packetSize);
    NSInteger size = [self.inputStream read:buffer maxLength:packetSize];
    if (size == packetSize) {
        NSData *data = [NSData dataWithBytesNoCopy:buffer length:size];
        if (self.dataBlock) {
            self.dataBlock(data, VEL_CURRENT_CMTIME);
        }
    } else {
        free(buffer);
    }
}

@end

File: Android/VeLiveQuickStartDemo/app/src/main/java/com/ttsdk/quickstart/helper/VeLiveFileReader.java
/**
 * Copyright (c) 2023 Beijing Volcano Engine Technology Ltd. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
package com.ttsdk.quickstart.helper;

import android.text.TextUtils;

import com.ss.avframework.utils.TimeUtils;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.nio.ByteBuffer;

public class VeLiveFileReader {
    private String mFilePath;
    private int mFrameSize;
    private Callback mCallback;
    private long mInterval;
    private boolean mEnable = false;
    private ByteBuffer mByteBuffer;

    public VeLiveFileReader() {
        mByteBuffer = ByteBuffer.allocateDirect(1920 * 1080 * 3 / 2); //  I420 一帧大小  
    }

    public interface Callback {
        void onByteBuffer(ByteBuffer byteBuffer, long pts);
    }

    private void setEnable(boolean enable) {
        mEnable = enable;
    }

    private boolean getEnable() {
        return mEnable;
    }

    public void start(String path, int frameSize, int interval, Callback callback) {
        if (TextUtils.isEmpty(path) || frameSize <= 0 || interval < 0 || callback == null) {
            return;
        }
        setEnable(true);
        mFilePath = path;
        mFrameSize = frameSize;
        mCallback = callback;
        mInterval = interval;
        new Thread(() -> {
            while (getEnable()) {
                doBusiness();
            }
        }).start();
    }

    public void stop() {
        setEnable(false);
    }

    private void doBusiness() {
        long timeUs = TimeUtils.nanoTime() / 1000;
        byte[] data = new byte[mFrameSize];
        long waitTimeUs = mInterval;
        FileInputStream dis = null;
        try {
            dis = new FileInputStream(mFilePath);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }

        while (mEnable) {
            try {
                int readCount = 0;
                while (readCount < mFrameSize) {
                    int tmp = dis.read(data, readCount, mFrameSize - readCount);
                    if (tmp >= 0) {
                        readCount += tmp;
                    } else {
                        break;
                    }
                }
                if (readCount < mFrameSize) {
                    break;
                }
                mByteBuffer.position(0);
                mByteBuffer.put(data);
                mByteBuffer.flip();
                mCallback.onByteBuffer(mByteBuffer, timeUs);
                timeUs += mInterval * 1000;
                waitTimeUs = timeUs - (TimeUtils.nanoTime() / 1000);
                if (waitTimeUs > 0) {
                    try {
                        Thread.sleep(waitTimeUs / 1000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            } catch (IOException e) {
                try {
                    dis.close();
                } catch (IOException ioException) {
                    ioException.printStackTrace();
                }
                break;
            }
        }
        if (dis != null) {
            try {
                dis.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
}
